crates/driver/src/domain/competition/quote.rs

```diff
@@ -0,0 +1,182 @@
+    pub sell_token: eth::TokenAddress,
+    pub buy_token: eth::TokenAddress,
+    pub amount: Amount,
+    pub valid_to: u32,
+    /// See [`crate::domain::competition::order::Partial`].
+    pub partial: bool,
+    pub quality: Quality,
+    pub gas_price: eth::EffectiveGasPrice,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Amount {
+    Sell(eth::U256),
+    Buy(eth::U256),
+}
+
+impl Order {
+    /// The asset being bought. Zero if this is a sell.
+    pub fn buy(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.buy_token,
+            },
+            Amount::Buy(amount) => eth::Asset {
+                amount,
+                token: self.buy_token,
+            },
+        }
+    }
+
+    /// The asset being sold. Zero if this is a buy.
+    pub fn sell(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(amount) => eth::Asset {
+                amount,
+                token: self.sell_token,
+            },
+            Amount::Buy(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.sell_token,
+            },
+        }
+    }
+
+    pub fn side(&self) -> competition::order::Side {
+        match self.amount {
+            Amount::Sell(..) => competition::order::Side::Sell,
+            Amount::Buy(..) => competition::order::Side::Buy,
+        }
+    }
+}
+
+/// Quality of the quote to be generated. This value determines the time
+/// allocated for the solver to generate the solution.
+#[derive(Debug, Clone, Copy)]
+pub enum Quality {
+    Fast,
+    Optimal,
+}
+
+impl Order {
+    /// Generate a quote for this order. This calls `/solve` on the solver with
+    /// a "fake" auction which contains a single order, and then determines
+    /// the quote for the order based on the solution that the solver
+    /// returns.
+    pub async fn quote(&self, solver: &Solver, config: &Config) -> Result<Quote, Error> {
+        let solution = solver
+            .solve(&self.fake_auction(), self.timeout(config))
+            .await?;
+        let fulfillment = solution
+            .trades
+            .into_iter()
+            .filter_map(|trade| match trade {
+                competition::solution::Trade::Fulfillment(fulfillment) => Some(fulfillment),
+                competition::solution::Trade::Jit(..) => None,
+            })
+            .exactly_one()
+            .map_err(|_| Error::QuotingFailed)?;
+        Ok(Quote::new(fulfillment, solution.interactions))
+    }
+
+    fn fake_auction(&self) -> competition::Auction {
+        competition::Auction {
+            id: None,
+            tokens: Default::default(),
+            orders: vec![competition::Order {
+                uid: Default::default(),
+                receiver: None,
+                valid_to: util::Timestamp::MAX,
+                sell: self.sell(),
+                buy: self.buy(),
```

Comment:
Looks like it got mixed up where `U256::MAX` and where `U256::ONE` should be returned.

Reply:
Not according to the code I pasted from the current http estimator implementation.
That says for sell orders `sell_amount = quote_amount, buy_amount = 1` and for buy orders `sell_amount = max, buy_amount = quote_amount`.
If the values are swapped it would be impossible for a solver to find a settlement that satisfies the fake trade. Selling some amount and expecting to buy `U256::MAX` is not achievable as is buying some amount while only selling `1`.
In order for the solver to find a valid solution and give surplus (which we interpret as the quote) we have to start from the opposite extreme point.